### PR TITLE
start design doc for stopping rule results

### LINKED ISSUE #124: design how to save atomic stopping rule results in simulations and report them
See discussion in #94 

To do:
Add design doc covering
- [ ] include current prototype using modified simulation function for reference / starting point
- [ ] how to unpack combined stopping rules such as `stopping = (A & B) | C | D` with atomic A, B, C, D 
- [ ] how to save atomic stopping rule results in simulation object
- [ ] how to summarize these

### PR DESCRIPTION
**Pull request by danielinteractive**
closes #124

### CODE REVIEW

File: design/design_stopping_sims.Rmd [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+number
+
+```
+
+# C. Count number of hits across all iterations and calculate return measures
+
+So, this is how we can get the order and names of the stopping rules for reporting later on. To count the frequency we have to get the information per iteration from the simulation method: 
+
+```{r, eval = F}
+stopLength <-length(object@stopping@stopList)
```
Comment (danielinteractive) [inline design/design_stopping_sims.Rmd:156]: This part I don't understand very well yet. I think if you just want the matrix of simulations in rows and the stopping rules in columns filled with `TRUE` and `FALSE` then we should try to get that / adapt the `stopTrial` method, as that one is already going through all the stopping rules and evaluating them.

**Review by danielinteractive — COMMENTED** [review]
thanks @robertadamsbayer , @ClaraBeckBayer and @dimitriskontosClinBAY for working on this!! Please see inside initial comments. Maybe we can discuss tomorrow more or in a dedicated session, let me know.

File: design/design_stopping_sims.Rmd [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+>>Stops (individual) for reason 3 StoppingMinPatients : 50 %  
+
+>>Stops (first hit) for reason 1 & 2 SStoppingMinCohorts and StoppingTargetProb : 75 %  
+>>Stops (first hit) for reason 3 StoppingMinPatients : 25 %  
+
+
+
+## Class and Method functions affected 
+
+* method "simulate"
```
Comment (danielinteractive) [inline design/design_stopping_sims.Rmd:225]: you can also here include prototype code by using another (similar) generic function name, e.g.:

```r
setMethod("new_simulate", ...)
```

File: design/design_stopping_sims.Rmd [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    # Total sample size for next dose is already >= 9.
+      Stopping_Rule5 <- StoppingPatientsNearDose(nPatients = 9,percentage = 0)
+or a combination (BOOLEAN logic) of mulptiple single crm-Pack stoping rule classes. 
+
+Additionally we would like to report :
+- How many times each individual stopping rule is triggered during the simulation (add up to more than 100 %).
+- Assuming the stopping rules used for the study simulation follow a pre-defined order of importance and assuming that multple stopping rules can be trigered at the
+   end of each simulation for example: 
+     Stopping_Rules_Combined <- (Stopping_Rule1 | Stopping_Rule2 | Stopping_Rule3  | Stopping_Rule4  | Stopping_Rule5)
+   only the most import of those trigered is coutned in each simulation run (add to 100 %).
```
[the remark this answers is not in the document]
Comment (danielinteractive) [reply design/design_stopping_sims.Rmd:32]: So after discussing this more in detail, it seems that we can just aggregate the final stopping results matrix (simulations x stopping rules with TRUE and FALSE entries) to obtain this summary. So the upstream code does not need to know about importance/ordering etc. and only the summary method / reporting part can take the importance/ordering from the user.

**Comment by danielinteractive** [comment]
Please let me know if you have questions on my comments, or whether it would be helpful to hack through this in a session together. I would like to move this forward. thanks!

File: design/design_stopping_sims.Rmd [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+   
+## Implementation suggestion
+
+# A. stopping rule naming for reporting
+  How to get the name of a stopping rule?
+  1. get the class name (might not be well readable)
+  2. should we have a slot in stopping rule classes (would require manual adjustment of all stopping rule classes) - 
+      requires to be present in all classes otherwise will throw errors.
```
[the remark this answers is not in the document]
Comment (danielinteractive) [reply design/design_stopping_sims.Rmd:40]: and for the `StoppingAny` class (and similarly for the `StoppingAll` class) you could something like this:

```r
setMethod( 
  f = "as.character", 
  signature = "StoppingAny", 
  definition = function(x, ...) { 
    ind_strings <- sapply(x@stopList, as.character)
    paste(ind_strings, collapse = " | ")
} )
```

File: design/design_stopping_sims.Rmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielinteractive) [reply design/design_stopping_sims.Rmd:32]: so maybe just being able to set a new argument to `TRUE` or `FALSE` for the summary/reporting methods in order to choose between 1) sums up to 100% vs. 2) sums up to more than 100% would be pragmatic solution here

File: R/Rules-methods.R [hunk 5]
```diff
@@ -1736,143 +1736,187 @@ setMethod("stopTrial",
...
 
+
+##' @describeIn as.character Convert stopping rules to character description
+##'
 setMethod(
   f = "as.character",
   signature = "StoppingMTDdistribution",
   definition = function(x, ...) {
-    paste("At least",
+    paste("\u2265",
```
Comment (danielinteractive) [inline R/Rules-methods.R:1834]: can you maybe just use the UTF8 character directly here? that would be easier to read

**Comment by danielinteractive** [comment]
@robertadamsbayer  @ClaraBeckBayer can you please resolve my comments above? = commenting and then resolve if cleared, or otherwise leave open. then it is easier for me to review the updates - see where are deltas compared to original proposals etc. thank you!

**Comment by ClaraBeckBayer** [comment]
@danielinteractive Hi Daniel, 
here some explanations for our commit: 
GetImageAttachment
Here you see what we have printed for individual results for StoppingAll und StoppingAny. However, when we browse the stopit object we can see in the console that we have the boolean values for all atomic stopping rules, it's just not "forwarded" to the further processing: 
GetImageAttachment (1)
GetImageAttachment (2)

**Comment by danielinteractive** [comment]
thanks @ClaraBeckBayer. I am not sure if I see the problem we discussed about. As far as I can see the individual results are returned, just not in a flat but rather a tree shape.

Consider the following example:

```r
# Create some data
data <- Data(x=c(0.1, 0.5, 1.5, 3, 6, 10, 10, 10),
             y=c(0, 0, 0, 0, 0, 0, 1, 0),
             cohort=c(0, 1, 2, 3, 4, 5, 5, 5),
             doseGrid=
               c(0.1, 0.5, 1.5, 3, 6,
                 seq(from=10, to=80, by=2)))

# Initialize the CRM model used to model the data
model <- LogisticLogNormal(mean=c(-0.85, 1),
                           cov=
                             matrix(c(1, -0.5, -0.5, 1),
                                    nrow=2),
                           ref_dose=56)

# Set-up some MCMC parameters and generate samples from the posterior
options <- McmcOptions(burnin=100,
                       step=2,
                       samples=2000)
set.seed(94)
samples <- mcmc(data, model, options)

# Define the rule for dose increments and calculate the maximum dose allowed
myIncrements <- IncrementsRelative(intervals=c(0, 20),
                                   increments=c(1, 0.33))
nextMaxDose <- maxDose(myIncrements,
                       data=data)

# Define the rule which will be used to select the next best dose
# based on the class 'NextBestNCRM'
myNextBest <- NextBestNCRM(target=c(0.2, 0.35),
                           overdose=c(0.35, 1),
                           maxOverdoseProb=0.25)

# Calculate the next best dose
doseRecommendation <- nextBest(myNextBest,
                               doselimit=nextMaxDose,
                               samples=samples, model=model, data=data)

# Define the stopping rules
myStopping1 <- StoppingMinCohorts(nCohorts=3)
myStopping2 <- StoppingTargetProb(target=c(0.2, 0.35),
                                  prob=0.5)
myStopping3 <- StoppingMinPatients(nPatients=20)
myStopping4 <- StoppingCohortsNearDose(2, 10)
myStopping5 <- StoppingMinCohorts(20)

# Combine the stopping rules, obtaining (in this specific example) a list of stopping
# rules of class 'StoppingAny'
myStopping <- (myStopping1 | myStopping2) & (myStopping3 | myStopping4) | myStopping5
#is(myStopping)
# Evaluate if to stop the Trial
result <- stopTrial(stopping=myStopping, dose=doseRecommendation$value,
          samples=samples, model=model, data=data)

individual_results_tree <- attr(result, "individual")

extract_individual <- function(tree) {
  if (is.list(tree)) {
    lapply(tree, flatten_tree)
  } else {
    ind_attr <- attr(tree, "individual")
    if (is.null(ind_attr)) {
      tree
    } else {
      lapply(ind_attr, flatten_tree)
    }
  }
}

ind_only_tree <- extract_individual(individual_results_tree)
ind_only_tree
unlist(ind_only_tree)
```

I deliberately made the stopping rule more complicated with 5 atomic stopping rules in it, using the Any and All combinations.
We get a complicated results tree back, but can further simplify it by recursing through the tree, extracting what we need and then unlisting that to flatten it. 

I would think that this can be used to obtain the flat results matrix of atomic stopping rules? But maybe I am missing something?

**Comment by danielinteractive** [comment]
working now on new branch, therefore need a new PR

**Pull request closed without merging**